homeassistant/components/recorder/statistics.py

```diff
@@ -1087,89 +1080,124 @@ def _reduce_statistics(
     return result
 
 
-def same_day(time1: datetime, time2: datetime) -> bool:
-    """Return True if time1 and time2 are in the same date."""
-    date1 = dt_util.as_local(time1).date()
-    date2 = dt_util.as_local(time2).date()
-    return date1 == date2
-
+def reduce_day_factory() -> (
+    tuple[
+        Callable[[datetime, datetime], bool],
+        Callable[[datetime], tuple[datetime, datetime]],
+    ]
+):
+    """Return functions to match same day and day start end."""
+    # We create _as_local_cached in the closure in case the timezone changes
+    _as_local_cached = lru_cache(maxsize=6)(dt_util.as_local)
```

Comment:
Since we call these in series the LRU allows us to avoid converting each previous day value for the next _same_day call so we only have to localize the time once.